src/goto-instrument/code_contracts.cpp

```diff
@@ -168,23 +169,56 @@ void code_contractst::add_quantified_variable(
   replace_symbolt &replace,
   irep_idt mode)
 {
-  // If the expression is a quantified expression, this function adds
-  // the quantified variable to the symbol table and to the expression map
-
-  // TODO Currently only works if the contract contains only a single
-  // quantified formula
-  // i.e. (1) the top-level element is a quantifier formula
-  // and (2) there are no inner quantifier formulae
-  // This TODO is handled in PR #5968
-
-  if(expression.id() == ID_exists || expression.id() == ID_forall)
+  if(expression.id() == ID_not || expression.id() == ID_typecast)
+  {
+    // For unary connectives, recursively check for
+    // nested quantified formulae in the term
+    unary_exprt unary_expression = to_unary_expr(expression);
+    exprt term = unary_expression.operands().at(0);
+    add_quantified_variable(term, replace, mode);
+  }
+  if(expression.id() == ID_notequal || expression.id() == ID_implies)
+  {
+    // For binary connectives, recursively check for
+    // nested quantified formulae in the left and right terms
+    binary_exprt binary_expression = to_binary_expr(expression);
+    exprt left = binary_expression.operands().at(0);
+    add_quantified_variable(left, replace, mode);
+    exprt right = binary_expression.operands().at(1);
+    add_quantified_variable(right, replace, mode);
+  }
+  if(expression.id() == ID_if)
+  {
+    // For ternary connectives, recursively check for
+    // nested quantified formulae in all three terms
+    if_exprt if_expression = to_if_expr(expression);
+    exprt condition = if_expression.operands().at(0);
+    add_quantified_variable(condition, replace, mode);
+    exprt first = if_expression.operands().at(1);
+    add_quantified_variable(first, replace, mode);
+    exprt second = if_expression.operands().at(2);
+    add_quantified_variable(second, replace, mode);
+  }
+  if(expression.id() == ID_and || expression.id() == ID_or)
+  {
+    // For multi-ary connectives, recursively check for
+    // nested quantified formulae in all terms
+    multi_ary_exprt multi_ary_expression = to_multi_ary_expr(expression);
+    for(auto operand : multi_ary_expression.operands())
+    {
+      add_quantified_variable(operand, replace, mode);
+    }
+  }
+  else if(expression.id() == ID_exists || expression.id() == ID_forall)
   {
-    // get quantified symbol
-    exprt tuple = expression.operands().front();
-    exprt quantified_variable = tuple.operands().front();
+    // When a quantifier expression is found,
+    // 1. get quantified symbol
+    quantifier_exprt quantifier_expression = to_quantifier_expr(expression);
+    exprt tuple = quantifier_expression.operands().at(0);
+    exprt quantified_variable = tuple.operands().at(0);
```

Comment:
I have made the modification as requested. However, I would like to point out that currently, it seems like quantifier expressions with multiple variables are not fully handled (or used, at least). This can be fixed in a separate PR.

Reply:
Yes, in my experience, most call sites use `.symbol()` in place of `.variables()` and expect a quantifier over a single variable. But thanks for the fix here. Looks good to me and we could add a regression test later once we handle multi-variable quantification properly.